### PR TITLE
New compdb

### PR DESCRIPTION
**Pull request by SteffenHeu**
refactored the compound database match to use the modular data model for easier integration of downstream processing

### CODE REVIEW

**Comment by SteffenHeu** [comment]
Double click on the PubChemId cell will automatically open the compound web page.
grafik

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        final Integer charge = molecularFormula.getCharge();
+        if (charge != null && charge != 0) {
+          logger.info(() -> "Compound " + annotation + " is not neutral as determined by smiles "
+              + MolecularFormulaManipulator.getString(molecularFormula) + ". charge = " + charge
+              + ". Adjusting protonation.");
+          MolecularFormulaManipulator.adjustProtonation(molecularFormula, -charge);
+        }
+        final double mass = MolecularFormulaManipulator.getMass(molecularFormula,
+            MolecularFormulaManipulator.MonoIsotopic);
+        return adduct.getMZ(mass);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java:128]: The whole smiles -> molecular formula -> neutral mass should maybe go in utility methods as its useful eitherway. Same for neutral mass from formula - I feel we might already have something for this

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/SimpleCompoundDBAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.stream.XMLStreamWriter;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Basic class for a compound annotation. The idea is not for it to be observable or so, but to
+ * carry a flexible amount of data while providing a set of minimum defined entries.
+ */
+public class SimpleCompoundDBAnnotation extends HashMap<DataType<?>, Object> implements
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/features/compoundannotations/SimpleCompoundDBAnnotation.java:45]: similar to ModularFeature etc I would internalize the HashMap and make this extend  ModularDataModel. Then you have more control over the data access with the <T> get methods

File: src/main/java/io/github/mzmine/datamodel/features/types/annotations/compounddb/CompoundAnnotationScoreType.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  @Override
+  public @NotNull String getUniqueID() {
+    return "compound_annotation_score";
+  }
+
+  @Override
+  public @NotNull String getHeaderString() {
+    return "Score";
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/features/types/annotations/compounddb/CompoundAnnotationScoreType.java:37]: If the header is just score you could also use the generic ScoreType. Specialized score types are only needed if you change the title or behavior. 
the columns will already read something like Compound match:Score

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java [hunk 4]
```diff
@@ -165,7 +212,10 @@ private void processOneLine(String values[], List<ImportType> linesWithIndices)
     var mobType = new MobilityType();
     var ccsType = new CCSType();
     var smilesType = new SmilesStructureType();
-    var adductType = new IonAdductType();
+    var adductType = new IonTypeType();
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java:215]: Maybe use DataType.get(IonTypeType.class) for all data types here

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java [hunk 5]
```diff
@@ -175,59 +225,36 @@ private void processOneLine(String values[], List<ImportType> linesWithIndices)
...
+        (entry.get(mobType) != null) ? Float.parseFloat(entry.get(mobType)) : null;
+    final Float lineCCS =
+        (entry.get(ccsType) != null) ? Float.parseFloat(entry.get(ccsType)) : null;
+    final Double neutralMass =
+        entry.get(neutralMassType) != null ? Double.parseDouble(entry.get(neutralMassType)) : null;
+    final String smiles = entry.get(smilesType);
+    final String pubchemId = entry.get(pubchemIdType);
+
+    CompoundDBAnnotation a = new SimpleCompoundDBAnnotation();
+    doIf(lineName != null, () -> a.put(compoundNameType, lineName));
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java:245]: This does not even save symbols :D but in turn you generate runnables 
I still like this pattern though. But maybe change it to
doIfNotNull(lineName, () -> ...))
or
putIfNotNull(key, value)

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java [hunk 6]
```diff
@@ -175,59 +225,36 @@ private void processOneLine(String values[], List<ImportType> linesWithIndices)
...
+    doIf(lineComment != null, () -> a.put(commentType, lineComment));
+    doIf(lineRT != null, () -> a.put(rtType, lineRT));
+    doIf(lineMob != null, () -> a.put(mobType, lineMob));
+    doIf(lineCCS != null, () -> a.put(ccsType, lineCCS));
+    doIf(smiles != null, () -> a.put(smilesType, smiles));
+    doIf(lineMZ != null, () -> a.put(MZType.class, lineMZ));
+    doIf(neutralMass != null, () -> a.put(neutralMassType, neutralMass));
+    doIf(IonType.parseFromString(lineAdduct) != null,
+        () -> a.put(ionTypeType, IonType.parseFromString(lineAdduct)));
+    doIf(pubchemId != null, () -> a.put(PubChemIdType.class, pubchemId));
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/LocalCSVDatabaseSearchTask.java:256]: maybe a bit confusing why you mix classes and data type objects here... but I know both work

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchTask.java [hunk 7]
```diff
@@ -46,7 +46,7 @@ public SpectralLibrarySearchTask(ParameterSet parameters, FeatureList[] featureL
     List<FeatureListRow> rows = new ArrayList<>();
     // add row type
     for (var flist : featureLists) {
-      flist.addRowType(DataTypes.get(SpectralLibraryMatchesType.class));
+      flist.addRowType(DataTypes.getInstance(SpectralLibraryMatchesType.class));
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchTask.java:49]: I actually liked the shorter version :D
DataTypes.get seems to tell me all I need to know

**Review by robinschmid — COMMENTED** [review]
small things

File: src/main/java/io/github/mzmine/datamodel/features/types/annotations/compounddb/CompoundAnnotationScoreType.java (hunk 3 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/types/annotations/compounddb/CompoundAnnotationScoreType.java:37]: ScoreType is abstract

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/SimpleCompoundDBAnnotation.java (hunk 2 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/compoundannotations/SimpleCompoundDBAnnotation.java:45]: doesnt the ModularDataModel rely on too much other stuff? Also you removed the ModularAnnotationType for a reason, or? 
I know this class itself almost reintroduces it, but not fully :D

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java (hunk 1 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java:128]: ok

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+  String XML_ELEMENT = "compound_db_annotation";
+  String XML_TYPE_ATTRIBUTE = "annotationtype";
+  String XML_NUM_ENTRIES_ATTR = "entries";
+
+  static List<CompoundDBAnnotation> buildCompoundsWithAddcuts(
+      CompoundDBAnnotation neutralAnnotation, IonNetworkLibrary library) {
+    final List<CompoundDBAnnotation> annotations = new ArrayList<>();
+    for (IonType adduct : library.getAllAdducts()) {
+      CompoundDBAnnotation clone = neutralAnnotation.clone();
```
[the remark this answers is not in the document]
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java:69]: sounds good

File: src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java (hunk 1 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/compoundannotations/CompoundDBAnnotation.java:128]: Some of the FormulaUtils stuff is really glitchy and i don't want to touch it out of fear of breaking somehting

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchTask.java (hunk 7 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchTask.java:49]: this one is more descriptive though

**Comment by robinschmid** [comment]
Perfect

**Pull request merged**